watch/watch.go

```diff
@@ -1,48 +1,61 @@
 package watch
 
 import (
+	"context"
 	"sync"
+	"time"
 
 	"github.com/docker/go-events"
+	"github.com/docker/swarmkit/watch/queue"
 )
 
-// dropErrClosed is a sink that suppresses ErrSinkClosed from Write, to avoid
-// debug log messages that may be confusing. It is possible that the queue
-// will try to write an event to its destination channel while the queue is
-// being removed from the broadcaster. Since the channel is closed before the
-// queue, there is a narrow window when this is possible. In some event-based
-// dropping events when a sink is removed from a broadcaster is a problem, but
-// for the usage in this watch package that's the expected behavior.
-type dropErrClosed struct {
-	sink events.Sink
-}
-
-func (s dropErrClosed) Write(event events.Event) error {
-	err := s.sink.Write(event)
-	if err == events.ErrSinkClosed {
-		return nil
-	}
-	return err
-}
-
-func (s dropErrClosed) Close() error {
-	return s.sink.Close()
+// ChannelSinkGenerator is a constructor of sinks that eventually lead to a
+// channel.
+type ChannelSinkGenerator interface {
+	NewChannelSink() (events.Sink, *events.Channel)
 }
 
 // Queue is the structure used to publish events and watch for them.
 type Queue struct {
+	sinkGen     ChannelSinkGenerator
+	limit       uint64
 	mu          sync.Mutex
 	broadcast   *events.Broadcaster
-	cancelFuncs map[*events.Channel]func()
+	cancelFuncs map[events.Sink]func()
+}
+
+// QueueOpts is a set of options that can be used to construct a new Queue.
+type QueueOpts struct {
+	ChannelSinkGenerator
+	QueueLimit uint64
 }
 
 // NewQueue creates a new publish/subscribe queue which supports watchers.
 // The channels that it will create for subscriptions will have the buffer
 // size specified by buffer.
 func NewQueue() *Queue {
+	return NewQueueWithOpts(QueueOpts{
+		ChannelSinkGenerator: &dropErrClosedChanGen{},
+		QueueLimit:           0,
+	})
+}
+
+// NewTimeoutLimitQueue creates a queue with a size limit and a request write
+// timeout.
+func NewTimeoutLimitQueue(timeout time.Duration, limit uint64) *Queue {
+	return NewQueueWithOpts(QueueOpts{
+		ChannelSinkGenerator: NewTimeoutDropErrSinkGen(timeout),
+		QueueLimit:           limit,
+	})
+}
+
+// NewQueueWithOpts creates a new Queue using the full set of options.
+func NewQueueWithOpts(opts QueueOpts) *Queue {
```

Comment:
To me this is begging for functional arguments: https://dave.cheney.net/2014/10/17/functional-options-for-friendly-apis